cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -400,9 +400,16 @@ public void programHistory(HttpRequest request, HttpResponder responder,
       throw new NotFoundException(String.format("Program history is not supported for program type '%s'.",
                                                 programType));
     }
-    long start = (startTs == null || startTs.isEmpty()) ? 0 : Long.parseLong(startTs);
-    long end = (endTs == null || endTs.isEmpty()) ? Long.MAX_VALUE : Long.parseLong(endTs);
-    getRuns(responder, Id.Program.from(namespaceId, appId, type, programId), status, start, end, resultLimit);
+      long start = (startTs == null || startTs.isEmpty()) ? 0 : Long.parseLong(startTs);
+      long end = (endTs == null || endTs.isEmpty()) ? Long.MAX_VALUE : Long.parseLong(endTs);
+
+      ProgramId id = Ids.namespace(namespaceId).app(appId).program(type, programId);
+      ProgramSpecification specification = lifecycleService.getProgramSpecification(id);
+      if (specification == null) {
+          throw new NotFoundException(id);
```

Comment:
For example if the request came in for a specific namespace:app:program, if the namespace itself doesn't exist, should be say the namespace is not found instead of saying the combination 'ns:app:prog' doesn't exist. I am ok with either but just wanted to make sure that option was considered.